File: src/android/src/com/christian/sentimask/SMEngine.java
package com.christian.sentimask;

import com.christian.sentimask.SMBase;
import com.christian.sentimask.SMProcessingDetails;

/**
 * SentiMask Engine
 */
public class SMEngine extends SMBase {

  private long nativeImageHandle;
  private long nativeDetailsHandle;

  public SMEngine() {
    this.SMEngineCreate();
    this.SMEngineInitialize();
  }

  public SMEngine(long handle) {
    super(handle);
  }

  public SMProcessingDetails processImage() {
    SMProcessingDetails pDetails = new SMProcessingDetails();
    boolean success = SMEngineProcessImage(0, pDetails.handle());
    if (success) return pDetails;
    return null;
  }

  public native boolean startProcessing(Object target);
  public native boolean stopProcessing();

  private native boolean SMEngineCreate();
  private native boolean SMEngineInitialize();
  private native boolean SMEngineProcessImage(long hImage, long hDetails);
  
}

File: www/SMEngine.js
const exec = require('cordova/exec');

const Config = require('./SMConfig');
const SMBase = require('./SMBase');
const SMProcessingDetails = require('./SMProcessingDetails');

class SMEngine extends SMBase {
  
  constructor() {
    exec((handle) => {
      super(handle);
    }, (error) => {
      throw xError;
    }, Config.name, 'SMEngineCreate', []);
  }
  
  processImage() {
    return new Promise((resolve, reject) => {
      exec((dHandle) => {
        let pDetails = new SMProcessingDetails(dHandle);
        resolve(pDetails);
      }, reject, Config.name, 'SMEngineProcessImage', [this.handle]);
    });
  }
  
  startProcessing() {
    return new Promise((resolve, reject) => {
      exec(resolve, reject, Config.name, 'SMEngineStartProcessing', [this.handle]);
    });
  }
  
  stopProcessing() {
    return new Promise((resolve, reject) => {
      exec(resolve, reject, Config.name, 'SMEngineStopProcessing', [this.handle]);
    });
  }
  
}

module.exports = SMEngine;
